cpp/src/decisiontree/decisiontree_impl.cuh

```diff
@@ -133,108 +134,75 @@ std::ostream &operator<<(std::ostream &os, const SparseTreeNode<T, L> &node) {
   return os;
 }
 
-template <typename T, typename L>
-struct Node_ID_info {
-  const SparseTreeNode<T, L> &node;
-  int unique_node_id;
-
-  Node_ID_info(const SparseTreeNode<T, L> &cfg_node, int cfg_unique_node_id)
-    : node(cfg_node), unique_node_id(cfg_unique_node_id) {}
-};
-
 template <class T, class L>
-void build_treelite_tree(TreeBuilderHandle tree_builder,
-                         DecisionTree::TreeMetaDataNode<T, L> *tree_ptr,
-                         int num_class) {
-  int node_id = 0;
-  TREELITE_CHECK(TreeliteTreeBuilderCreateNode(tree_builder, node_id));
-
-  std::queue<Node_ID_info<T, L>> cur_level_queue;
-  std::queue<Node_ID_info<T, L>> next_level_queue;
-
-  cur_level_queue.push(Node_ID_info<T, L>(tree_ptr->sparsetree[0], 0));
-  node_id = -1;
-
-  while (!cur_level_queue.empty()) {
-    int cur_level_size = cur_level_queue.size();
-    node_id += cur_level_size;
-
-    for (int i = 0; i < cur_level_size; i++) {
-      Node_ID_info<T, L> q_node = cur_level_queue.front();
-      cur_level_queue.pop();
-
-      bool is_leaf_node = q_node.node.colid == -1;
+tl::Tree<T, T> build_treelite_tree(
```

Comment:
Is it fair to assume that the thresholds and leaf outputs have the same type? The template argument `L` is not used anywhere.

Reply:
The threshold and leaf output types will always be the same because of Treelite's static assertions [here](https://github.com/dmlc/treelite/blob/mainline/include/treelite/tree.h#L247-L260) and the fact that we don't use `uint32_t` for our leaf output type anywhere. The input leaf type (L) may be different from the threshold type, however. Specifically, it can be an int (lines 521-530) which gets cast to either a double or a float at line 190. We never reference L, but we take advantage of the fact that T and L can be different.

Reply:
Got it. We can keep it for now. However, in the future, we may want to change the representation for the categorical leaf output, from one-hot encoding to a single integer.